### PR TITLE
Format tweaks to devel branch, try 2

### PR DESCRIPTION
**Pull request by dnystrom1**
Not sure why there were merge conflicts before because I did not get any conflicts after pulling the new lanl/vpic devel branch into my dnystrom1/vpic devel branch and then pulling my dnystrom1/devel_format_tweaks branch into my dnystrom1/vpic devel branch.  However, here is my second try, this time from my dnystrom1/vpic devel branch and hopefully this will work.

### CODE REVIEW

File: src/field_advance/standard/pipeline/compute_div_b_err_pipeline.h [hunk 1]
```diff
@@ -13,7 +13,7 @@ typedef struct pipeline_args
   const grid_t *              g;
 } pipeline_args_t;
 
-#define f(x,y,z) f[ VOXEL(x,y,z, nx,ny,nz) ]
+#define f(x,y,z) f[ VOXEL( x, y, z, nx, ny, nz ) ]
```
Comment (rfbird) [inline src/field_advance/standard/pipeline/compute_div_b_err_pipeline.h]: Maybe it's time to define a style guide. I think this should be a third option:

` VOXEL(x, y, z, nx, ny, nz)`

Comment (rfbird) [reply src/field_advance/standard/pipeline/compute_div_b_err_pipeline.h]: After consulting clang format, which I want to move us to eventually, I'm fine with how you have it!

**Comment by dnystrom1** [comment]
Is clang format something I should read and learn about?  If so, can you point me to a reference or link?

**Comment by rfbird** [comment]
> Is clang format something I should read and learn about? If so, can you point me to a reference or link?

Yes, it's probably worth while. It just gives your an automated way to format the code based on some rules, and we can even have the option of having it format the code when you push. 

You can read about the basics [here](https://clang.llvm.org/docs/ClangFormat.html) and [here](https://clang.llvm.org/docs/ClangFormatStyleOptions.html). Once I've had chance to think about how I'd like it to work we can discuss further (my other projects uses it very successfully, I'll steal their workflow..)

**Comment by dnystrom1** [comment]
Commenting on your first set of comments that I just saw and read.  A couple of quotes come to mind:-).

"Beauty is in the eyes of the beholder"

and

"There is no accounting for taste."

I took a look at what clang_format did to the format of source files you processed with it and was somewhat dismayed.  It seemed that less than half of the changed lines of code were formatted well.  Can clang_format be customized?  Also, can I manually process a file or set of files with clang_format?  I think it would be helpful for me to do that.

I think it would be a fine idea to remove the C-style casts.  Would you want to do that before or after the release?  I would think it would be better after the release because I can imagine the need for thorough testing.

On a different note, I am in a Floating Point Analysis training session all of this morning.  Was in an HPCToolkit training session yesterday.  Anyway, I'm thinking I will learn about some useful tools today that would be nice to run vpic through.

**Comment by dnystrom1** [comment]
OK, looking at your links above for documentation, I see that those should answer some of my questions.  So, I will start reading and exploring this.

BTW, beyond aesthetic considerations, a lot of the files that I reformat are done for the purpose of being able to compare various files visually using a tool like xxdiff or meld so I can study code differences.  For instance, right now, I can compare all of the ivpic changes from Guangye to the current devel branch and see the real changes since I have eliminated format differences.

**Comment by rfbird** [comment]
What do you mean by "formatted well"? Can you give me a link to an example? Sounds like an eye of the beholder thing to me, I don't see anything *too* offensive ... 

To run manually, install clang-format and then run `make format` if you have the branch install, or you can manually run by using `clang-format -style=file <src_filename_here>`. Make sure you run that from a folder that has our [proposed] [clang format file](https://github.com/lanl/vpic/blob/214030d0511e6c033de5d4b3770983227e51a1ca/.clang-format)

Casts I think we can do either before or after, perhaps naively, I trust the C++ style casts. Let's wait till after I guess

I wanted to go the FP workshop, but they wouldn't broad bast it. Nigel is looking at FP16 for VPIC so it's very timely. I'm interested to hear what you learn

**Comment by dnystrom1** [comment]
Let me experiment a bit before trying to provide some more feedback.  Looks like I should be able to run the current devel branch through clang-format and then xxdiff all the files which will then make it a lot easier for me to do a good evaluation and provide useful feedback.

How long do you think it would take to change C-style casts to C++?  I'm fine with doing it before the release rather than after as long as there is a commitment to allowing time for adequate testing even if it slips the release date.

What about the possible adoption of clang-format?  Is that something you would like to consider for this release?  I'm open to that as well i.e. I'm willing to commit to taking the time to do the evaluation and learn how to use it sooner rather than later.

**Comment by rfbird** [comment]
I'm happy for clang format to go in the release, yea. I'd rather have that than the casts. 

Let's see how far you get with your evaluation and go from there. If possible, I'm not looking to change the config file too much as right now it matches my ECP project

**Comment by dnystrom1** [comment]
I think I made some good progress yesterday and last night on exploring clang-format.  I am feeling more encouraged than I was before reading the documentation on clang-format and experimenting with it.  I still have some more experimenting to do.

Can you help me understand what you want to achieve by moving to the use of a tool like clang-format?  Understanding what you want to achieve would be helpful to me as I continue with and finish up my experimenting.

**Comment by rfbird** [comment]
Sure, the motivation is pretty simple: have the code formatting one consistent way, and keep it that way.

Right now we have a lot of Kevin style, and a little bit of both Bob and Dave style. It looked like your most recent refactoring PRs might have taken an hour or two to do, but with the tool that's now instant and guarantees consistency. 

We can even have Github complain when someone makes a PR that does not meet the "standard"

**Comment by dnystrom1** [comment]
Okay.  Can you describe the anticipated workflow?  I was assuming that when someone submitted a PR, GitHub would just reformat it as part of the process of merging in the changes rather than complaining.  What would a developer need to do in order to make sure they were submitting a PR that met the format standard?

For instance, for my experimenting, I have a copy of the source code where I have changed the names of the .h and .cc files to be .orig.h and .orig.cc.  Then, I have a script that runs "clang-format -style=file xxx.orig.cc >& xxx.cc", etc.  I can easily make a better script where the script moves .h and .cc files to some temporary filename, runs clang-format on these temporary files to generate new, format compliant .h and .cc files and then removes the temporary files.  I could then run such a script manually before doing commits.  If I knew how, I could just make such a script a part of the commit process.  I assume you know how to do that.

Anyway, how do you imagine this working as part of the developers workflow?

On a different note, how robust is all this against an evolving clang-format?  For instance, I am using clang-9.0 on Darwin but there are choices.  And, my Linux laptop is running Fedora 28 and when I checked what version of clang was available in the Fedora rpm repo, it only has 6.0.  In reviewing the documentation, I see that configure options can evolve and become deprecated.  So, I wonder if compliance with the VPIC format standard will require access to specific versions of clang-format and whether for a constant input, the output of clang-format will change over time.  I could imagine it would, especially since C++ is an evolving language, and I wonder how we would deal with that.  Have you thought about things like this?  I'm not suggesting that this is an impediment to using something like clang-format, only that it makes sense to consider these possibilities and think about how to deal with them.

**Comment by dnystrom1** [comment]
BTW, sent you email on a different topic.  Not sure how frequently you check your LANL email.

**Comment by rfbird** [comment]
> Okay. Can you describe the anticipated workflow? I was assuming that when someone submitted a PR, GitHub would just reformat it as part of the process of merging in the changes rather than complaining. What would a developer need to do in order to make sure they were submitting a PR that met the format standard?
> 
> For instance, for my experimenting, I have a copy of the source code where I have changed the names of the .h and .cc files to be .orig.h and .orig.cc. Then, I have a script that runs "clang-format -style=file xxx.orig.cc >& xxx.cc", etc. I can easily make a better script where the script moves .h and .cc files to some temporary filename, runs clang-format on these temporary files to generate new, format compliant .h and .cc files and then removes the temporary files. I could then run such a script manually before doing commits. If I knew how, I could just make such a script a part of the commit process. I assume you know how to do that.
> 
> Anyway, how do you imagine this working as part of the developers workflow?
> 
> On a different note, how robust is all this against an evolving clang-format? For instance, I am using clang-9.0 on Darwin but there are choices. And, my Linux laptop is running Fedora 28 and when I checked what version of clang was available in the Fedora rpm repo, it only has 6.0. In reviewing the documentation, I see that configure options can evolve and become deprecated. So, I wonder if compliance with the VPIC format standard will require access to specific versions of clang-format and whether for a constant input, the output of clang-format will change over time. I could imagine it would, especially since C++ is an evolving language, and I wonder how we would deal with that. Have you thought about things like this? I'm not suggesting that this is an impediment to using something like clang-format, only that it makes sense to consider these possibilities and think about how to deal with them.

Workflow (which, if we accept it I will document somewhere sensible such as the readme):

1) Make edits to code
2) run `make format`
3) commit and push

(no scripts required)

I think it's pretty robust. We have a format file with rules, so we have full control. To my knowledge they get applied consistently across versions, but if they don't we can always update our rule file. The most obvious example would be if default behavior changes in a newer version, we can then decide which one we want to adopt and either a) update the code to use that, or b) update our rule file to specify it should use the "old" behaviour. I don't see any major failure-case, do you?

Per before, the reason I'm confident in this is that it's adopted by my ECP project and has been shown to work well. If it's good enough for that, it's good enough for us? Also, I'd argue even if there are small kinks, it would still give us a more consistent style overall than we have no.

Finally, it would avoid any wasted time manually reformatting code :). I'm still very keen to move to a more open development model, where we have more contributions from external users, and I think this will help set expectations.

Finally, we'll have new people working regularly on VPIC contributions (mostly on the GPU stuff), I think that puts us at/above the critical mass (>= 3 regular contributors) where we really need a style guide

**Pull request merged**